File: modules/core/app/model/CorrectionException.java
package model;

public class CorrectionException extends Exception {
  
  private static final long serialVersionUID = -639146851013219568L;
  
  private final String learnerSolution;
  
  public CorrectionException(String theLearnerSolution, String msg) {
    super(msg);
    learnerSolution = theLearnerSolution;
  }
  
  public CorrectionException(String theLearnerSolution, String msg, Throwable cause) {
    super(msg, cause);
    learnerSolution = theLearnerSolution;
  }
  
  public CorrectionException(String msg, Throwable cause) {
    this("", msg, cause);
  }
  
  public String getLearnerSolution() {
    return learnerSolution;
  }
  
}

File: app/model/core/CorrectionException.scala
package model.core

sealed abstract class CorrectionException(msg: String) extends Exception(msg)

final case class NoSuchExerciseException(id: Int) extends CorrectionException(s"Es gibt keine Aufgabe mit der ID '$id'")

final case class NoSuchPartException(partStr: String) extends CorrectionException(s"Es gibt keinen Aufgabenteil mit dem Namen '$partStr'")

object SolutionTransferException extends CorrectionException("Es gab einen Fehler bei der Übertragung ihrer Lösung!")

final case class OtherCorrectionException(cause: Throwable) extends CorrectionException(cause.getMessage)